source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,340 @@
+
+  MessagePtr request(new RequestMessageImpl());
+  request->headers().insertContentType().value().setReference(
+      Headers::get().ContentTypeValues.Json);
+  request->headers().insertPath().value().setReference(POST_ATTACHMENT_PATH);
+  request->headers().insertHost().value().setReference(SERVER_AUTHORITY);
+  request->headers().insertMethod().value().setReference(Headers::get().MethodValues.Post);
+  request->body().reset(new Buffer::OwnedImpl(config_->attachmentJson()));
+
+  state_ = State::CreateConfig;
+  in_flight_request_ = cm_.httpAsyncClientForCluster(config_->clusterName())
+                           .send(std::move(request), *this, config_->requestTimeout());
+
+  if (in_flight_request_ == nullptr) {
+    ENVOY_LOG(info, "Squash: can't create request for squash server");
+    state_ = State::Initial;
+    return FilterHeadersStatus::Continue;
+  }
+
+  attachment_timeout_timer_ =
+      decoder_callbacks_->dispatcher().createTimer([this]() -> void { doneSquashing(); });
+  attachment_timeout_timer_->enableTimer(config_->attachmentTimeout());
+  // check if the timer expired inline.
+  if (state_ == State::Initial) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterDataStatus SquashFilter::decodeData(Buffer::Instance&, bool) {
+  if (state_ == State::Initial) {
+    return FilterDataStatus::Continue;
+  } else {
+    return FilterDataStatus::StopIterationAndBuffer;
+  }
+}
+
+FilterTrailersStatus SquashFilter::decodeTrailers(HeaderMap&) {
+  if (state_ == State::Initial) {
+    return FilterTrailersStatus::Continue;
+  } else {
+    return FilterTrailersStatus::StopIteration;
+  }
+}
+
+void SquashFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+void SquashFilter::onSuccess(MessagePtr&& m) {
+  in_flight_request_ = nullptr;
+
+  switch (state_) {
+
+  case State::Initial: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send success callback when no request is in progress");
+    break;
+  }
+  case State::CreateConfig: {
+    // get the config object that was created
+    if (Utility::getResponseStatus(m->headers()) != enumToInt(Code::Created)) {
+      ENVOY_LOG(info, "Squash: can't create attachment object. status {} - not squashing",
+                m->headers().Status()->value().c_str());
+      doneSquashing();
+    } else {
+      state_ = State::CheckAttachment;
+
+      std::string debugAttachmentId;
+      try {
+        Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+        debugAttachmentId =
+            json_config->getObject("metadata", true)->getString("name", EMPTY_STRING);
+      } catch (Json::Exception&) {
+        debugAttachmentId = EMPTY_STRING;
+      }
+
+      if (debugAttachmentId.empty()) {
+        ENVOY_LOG(info, "Squash: failed to parse debug attachment object - check server settings.");
```

Comment:
Do you think these failure modes are worth tracking as counters?

Reply:
Not sure as I don't expect squash invocations to happen unattended. The main use case (at least for now) as a human who wants to debug a request that is traveling through the mesh.

Reply:
Not sure as I don't expect squash invocations to happen unattended. The main use case (at least for now) as a human who wants to debug a request that is traveling through the mesh.